src/coreclr/gc/env/gcenv.object.h

```diff
@@ -127,7 +131,15 @@ class MethodTable
 
     bool HasCriticalFinalizer()
     {
-        return (m_flags & MTFlag_HasCriticalFinalizer) != 0;
+#ifdef BUILD_AS_STANDALONE
+        if (g_oldMethodTableFlags)
+        {
+            // This flag is used for .NET 8 or below
+            const int Old_MTFlag_HasCriticalFinalizer = 0x08000000;
+            return (m_flags & Old_MTFlag_HasCriticalFinalizer) != 0;
+        }
+#endif
+        return (m_flags & MTFlag_HasCriticalFinalizer) && !HasComponentSize();
```

Comment:
>return (m_flags & MTFlag_HasCriticalFinalizer) && !HasComponentSize(); [](http://example.com/codeflow?start=8&length=70)

nit - it'd be a bit better to check for the array type first just to make it clear that if it's an array it cannot have a finalizer